### PR TITLE
models unit testing

### PR DESCRIPTION
**Pull request by bassantmedhat**
liver, lesion and lobes models unit testing

### CODE REVIEW

File: tests/test_liver.py [hunk 1]
```diff
@@ -0,0 +1,372 @@
...
+        mode=config.dataset["mode"],
+        shuffle=config.training["shuffle"],
+    )
+    test_dataload = test_dataloader.get_testing_data()
+    test_loss, test_metric = liver_object_sw.test_sliding_window(
+        dataloader=test_dataload
+    )
+
+    assert test_loss <= 1
+    assert test_metric <= 0.5
```
Comment (Muhammad-Al-Barbary) [inline tests/test_liver.py:301]: This doesn't guarantee that the code is logically bug free. Calculate the expected values and save them as references.

File: tests/test_liver.py [hunk 2]
```diff
@@ -0,0 +1,372 @@
...
+
+
+@pytest.mark.parametrize(
```
Comment (Muhammad-Al-Barbary) [inline tests/test_liver.py:304]: Why do you parametrize them if you use a single value per parameter? I think you can utilize this parameterization technique to generalize the tests and reduce redundancy as mentioned above.

File: tests/test_liver.py [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+    assert (
+        liver_prediction.shape[2:]
+        == nib.load(config.test["test_volume"]).get_fdata().shape
+    )  # original volume dimension
+
+    prediction = liver_prediction.cpu()
+    prediction = prediction.numpy()
+    final_path = path
+    true = np.load(final_path)
+    assert np.allclose(prediction, true, 0.01)
```
Comment (Muhammad-Al-Barbary) [inline tests/test_liver.py:340]: This is what I meant regarding the per-pixel comparison.

File: tests/test_lesion.py [hunk 4]
```diff
@@ -0,0 +1,211 @@
...
+
+    liver_mask = segment_liver(
+        prediction_path=config.test["test_volume"],
+        modality="CT",
+        inference="sliding_window",
+        cp_path=config.test["reference_sliding_window"],
+    )
+    lesion_inference = "3D"
+    volume_dir = "tests/testdata/resized_liver/volume/resized_liver.nii"
+    close = MorphologicalClosing(iters=4)
```
Comment (Muhammad-Al-Barbary) [inline tests/test_lesion.py:139]: This is not necessary in the test, it boosts the segmentation performance, but this is not the test scope.

File: tests/test_lobe.py [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+    """
+
+    liver_prediction = segment_liver(
+        prediction_path=config.test["test_volume"],
+        modality="CT",
+        inference="sliding_window",
+        cp_path=config.test["reference_sliding_window"],
+    )
+    lobe_inference = "3D"
+    extract = Compose([Dilation(), ConvexHull()])
```
Comment (Muhammad-Al-Barbary) [inline tests/test_lobe.py:239]: same here, not really necessary.

**Review by Muhammad-Al-Barbary — COMMENTED** [review]
Great work on the styling convention, please make sure to agree on it with all the team members, so that they stick to it as well.
Please apply the liver tests comments to those of the lesions and the lobes.

File: tests/test_lesion.py (hunk 4 above)
Comment (bassantmedhat) [reply tests/test_lesion.py:139]: I made this to  adjust the output dimensions of the liver mask after performing segmentation.

Comment (Muhammad-Al-Barbary) [reply tests/test_lesion.py:139]: I meant the Morphological Closing step.

File: tests/test_lesion.py [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+    prediction = lesion_obj.predict_2dto3d(
+        volume_path=volume_dir,
+        liver_mask=close(liver_mask[0]).permute(3, 0, 1, 2)
+        if lesion_inference == "3D"
```
Comment (Muhammad-Al-Barbary) [inline tests/test_lesion.py:143]: you can remove the unnecessary if conditions, since lesion_inference is hardcoded anyways.

File: tests/test_liver.py (hunk 1 above)
Comment (bassantmedhat) [reply tests/test_liver.py:301]: but the numbers change every time but they still within the same range.

Comment (Muhammad-Al-Barbary) [reply tests/test_liver.py]: if you set seeds and use the same test cases, the values are supposed to be deterministic (maybe with minor variations as a result of floating point operations).

File: tests/test_liver.py [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+
+@pytest.fixture
+def liver_obj():
```
[the remark this answers is not in the document]
Comment (bassantmedhat) [reply tests/test_liver.py:226]: it is used in the next function.

File: tests/test_liver.py [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+    assert liver_prediction.shape[0] == 1
+    assert liver_prediction.shape[1] == 1  # number of channels
+    assert (
+        liver_prediction.shape[2:]
+        == nib.load(config.test["test_volume"]).get_fdata().shape
+    )  # original volume dimension
+
+    prediction = liver_prediction.cpu()
+    prediction = prediction.numpy()
+    final_path = path
```
[the remark this answers is not in the document]
Comment (bassantmedhat) [reply tests/test_liver.py:220]: it is not duplication, the process is done in two steps for readability.

File: tests/test_liver.py [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+    assert torch.max(prediction) <= 1
+
+
+@pytest.fixture
+def liver_object_sw():
+    liver_object_sw = LiverSegmentation(modality="CT", inference="sliding_window")
+    return liver_object_sw
+
+
+def test_predict_sliding_window(liver_object_sw):
```
[the remark this answers is not in the document]
Comment (bassantmedhat) [reply tests/test_liver.py:257]: The two functions test different prediction functions. If I parametrize them, it will lead to code repetition in the assertions. Should I merge them?

File: tests/test_lesion.py [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+    transform_name (string): "2d_ct_transform".
+    lesion_obj (object): The lesion object that provides the get_pretraining_transforms method.
+    ----------
+    """
+
+    output_transform = lesion_obj.get_pretraining_transforms(transform_name)
+    assert isinstance(output_transform, Compose)
+
+    if transform_name == "2d_ct_transform":
+        expected_transfom = [
```
Comment (Muhammad-Al-Barbary) [inline tests/test_lesion.py:55]: what if you change the transforms in the lesions module? should the test fail then? maybe this test should just check that a compose object containing dictionary transforms is returned (by checking that each of them is an instance of MapTransform, for example)

File: tests/test_lesion.py [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+    liver_mask = segment_liver(
+        prediction_path=config.test["test_volume"],
+        modality="CT",
+        inference="sliding_window",
+        cp_path=config.test["reference_sliding_window"],
+    )
+    lesion_inference = "3D"
+    volume_dir = config.test["test_volume"]
+    close = MorphologicalClosing(iters=4)
+    prediction = lesion_obj.predict_2dto3d(
```
Comment (Muhammad-Al-Barbary) [inline tests/test_lesion.py:140]: why aren't you checking the values of the prediction?

File: tests/test_lesion.py [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+        batch_callback_epochs=100,
+        save_weight=True,
+        save_path=config.test["lesion_cp"],
+        test_batch_callback=False,
+    )
+    model.load_checkpoint(config.test["lesion_cp"])
+    loaded_weights = model.network.state_dict()
+    # Check that weights match
+    for i in init_weights.keys():
+        assert torch.allclose(init_weights[i], loaded_weights[i])
```
Comment (Muhammad-Al-Barbary) [inline tests/test_lesion.py:211]: variables names are misleading. change them into reference_weights and trained_weights, for example.

File: tests/test_liver.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Muhammad-Al-Barbary) [reply tests/test_liver.py:220]: final_path = path ?

File: tests/test_liver.py [hunk 13]
```diff
@@ -0,0 +1,370 @@
...
+            FillHolesD,
+            KeepLargestConnectedComponentD,
+        ]
+
+    for t, e in zip(output_transform.transforms, expected_transfom):
+        assert isinstance(t, e)
+
+
+
+def test_segment_liver_2d(modality="CT", inference= "3D", path= "tests/testdata/testcases/predicted_3d.npy"):
```
Comment (Muhammad-Al-Barbary) [inline tests/test_liver.py:189]: why is this test taking direct arguments?

File: tests/test_liver.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Muhammad-Al-Barbary) [reply tests/test_liver.py:226]: it is used in the above ones as well, isn't it?

File: tests/test_liver.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Muhammad-Al-Barbary) [reply tests/test_liver.py:257]: just parametrize the prediction methods and the reference paths, then the same assertions can be run for each pair of them.

File: tests/test_liver.py [hunk 14]
```diff
@@ -0,0 +1,370 @@
...
+        test_size=1,
+        mode= "2D",
+        shuffle=config.training["shuffle"],
+    )
+    test_dataload = test_dataloader.get_testing_data()
+    test_loss, test_metric = liver_object_sw.test_sliding_window(
+        dataloader=test_dataload
+    )
+
+    assert round(test_loss,1) == 0.9
```
Comment (Muhammad-Al-Barbary) [inline tests/test_liver.py:298]: I think the specified number of decimals to round to is very small, increase it if possible to increase test accuracy.

File: tests/test_liver.py [hunk 15]
```diff
@@ -0,0 +1,370 @@
...
+        save_weight=True,
+        save_path=config.test["liver_cp_sw"],
+        test_batch_callback=False,
+    )
+    model.load_checkpoint(config.test["liver_cp_sw"])
+    loaded_weights = model.network.state_dict()
+
+    # Check that weights match
+    for i in init_weights.keys():
+        assert torch.allclose(init_weights[i], loaded_weights[i])
```
Comment (Muhammad-Al-Barbary) [inline tests/test_liver.py:368]: same comment for the variables names

File: tests/test_lobe.py
Comment (Muhammad-Al-Barbary) [inline tests/test_lobe.py:1]: apply previous comments here as well.

**Review by Muhammad-Al-Barbary — COMMENTED** [review]
You have several unaddressed comments from last review, please address them as soon as possible. Feel free to discuss any of them.

File: tests/testdata/reference_lesion_cp
[the remark this answers is not in the document]
Comment (bassantmedhat) [reply tests/testdata/reference_lesion_cp:1]: done

File: tests/testdata/Temporary lessions/Test/mask/resized_mask_0.png
[the remark this answers is not in the document]
Comment (bassantmedhat) [reply tests/testdata/Temporary lessions/Test/mask/resized_mask_0.png:1]: In the lesion model, PNG files are required for training, so I have sliced them from a small input volume.

File: tests/test_liver.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (bassantmedhat) [reply tests/test_liver.py:226]: okay, got it

[the remark this answers is not in the document]
Comment (bassantmedhat) [reply tests/test_liver.py:226]: do you mean above the first function using it  ?

File: tests/test_liver.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (bassantmedhat) [reply tests/test_liver.py:220]: done

File: tests/test_liver.py (hunk 15 above)
Comment (bassantmedhat) [reply tests/test_liver.py:368]: done

File: tests/test_lesion.py (hunk 10 above)
Comment (bassantmedhat) [reply tests/test_lesion.py]: Is it necessary to also check the types of transforms?

File: tests/test_lesion.py (hunk 11 above)
Comment (bassantmedhat) [reply tests/test_lesion.py:139]: done

File: tests/test_lesion.py (hunk 12 above)
Comment (bassantmedhat) [reply tests/test_lesion.py:210]: done

File: tests/test_lesion.py [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+    AsDiscreteD,
+    FillHolesD,
+    ScaleIntensityRanged,
+    RemoveSmallObjectsD,
+)
+
+
+@pytest.fixture
+def lesion_obj():
+    lesion_obj = LesionSegmentation()
```
Comment (Muhammad-Al-Barbary) [inline tests/test_lesion.py:34]: why aren't you setting a seed here?

File: tests/test_lesion.py [hunk 17]
```diff
@@ -0,0 +1,215 @@
...
+    return lesion_obj
+
+def test_get_pretraining_transforms(lesion_obj):
+    """
+    Tests get_pretraining_transforms function.
+    verifies the functionality by comparing the output transforms with the expected transforms based on the provided transform_name.
+
+    Parameters:
```
Comment (Muhammad-Al-Barbary) [inline tests/test_lesion.py:42]: what parameters?

File: tests/test_lesion.py [hunk 18]
```diff
@@ -0,0 +1,215 @@
...
+
+
+
+
+def test_get_pretesting_transforms(lesion_obj):
+    """
+    Tests the get_pretesting_transforms function.
+    verifies the functionality by comparing the output transforms with the expected transforms based on the provided transform_name.
+
+    Parameters:
```
Comment (Muhammad-Al-Barbary) [inline tests/test_lesion.py:76]: same as above, apply it to all tests

File: tests/test_lesion.py [hunk 19]
```diff
@@ -0,0 +1,215 @@
...
+
+    )
+
+
+    assert isinstance(prediction, torch.Tensor)
+    assert prediction.shape[0] == 1
+    assert prediction.shape[1] == 1  # number of channels
+
+    assert (
+        prediction.shape[2:] == nib.load(config.test["test_volume"]).get_fdata().shape
```
Comment (Muhammad-Al-Barbary) [inline tests/test_lesion.py:154]: the == operation might cause trouble because of the floating point data type. Maybe you should replace it with an "is close" assertion.

File: tests/test_lesion.py [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+
+def test_segment_lesion():
+    """
+    Tests segment_lesion function (lesion_inference = '3D').
+    verifies the functionality by performing segmentation using a 3D volume (with size 64,64,...)
+    Compares the results with the true reference segmentation.
+    """
+    lesion_prediction = segment_lesion(
+        prediction_path=config.test["test_volume"],
+        liver_inference="sliding_window",
```
Comment (Muhammad-Al-Barbary) [inline tests/test_lesion.py:167]: Sliding window technique using 3d models is usually less efficient. Try using a 2D model for liver prediction since it's not the one being tested, and compare efficiency. Just a suggestion, put it on your to-do list.

File: tests/test_lesion.py [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+    assert (
+        lesion_prediction.shape[2:]
+        == nib.load(config.test["test_volume"]).get_fdata().shape
+    )  # original volume dimension
+
+    lesion_prediction = lesion_prediction.cpu()
+    lesion_prediction = lesion_prediction.numpy()
+
+
+    ref_path = "tests/testdata/testcases/predicted_array_lesion.npy"
```
Comment (Muhammad-Al-Barbary) [inline tests/test_lesion.py:186]: doesn't look clean. Put it in the configs.

File: tests/test_lesion.py [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+
+def test_train():
+    """
+    Tests train function (epoch: 1 , inference = '3D', volume = '64,64,..' ) .
+    Compares the weights of the resulted checkpoint with the reference checkpoint's weights.
+    """
+    model = LesionSegmentation(inference="3D")
+    # load reference checkpoint of training 1 epoch on a volume
+    model.load_checkpoint(config.test["reference_lesion_cp"])
+    reference_weights = model.network.state_dict()
```
Comment (Muhammad-Al-Barbary) [inline tests/test_lesion.py:199]: I think it will be more understandable if you move these two lines (loading reference cp) below loading the training cp, because it seems as if 'model' is what is being trained. This is also a suggestion.

File: tests/test_liver.py (hunk 13 above)
Comment (Muhammad-Al-Barbary) [reply tests/test_liver.py:198]: ??

File: tests/test_liver.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (Muhammad-Al-Barbary) [reply tests/test_liver.py:229]: ??

File: tests/test_liver.py [hunk 23]
```diff
@@ -0,0 +1,362 @@
...
+        pin_memory=False,
+        test_size=1,
+        mode= "2D",
+        shuffle=config.training["shuffle"],
+    )
+    test_dataload = test_dataloader.get_testing_data()
+    test_loss, test_metric = liver_object_sw.test_sliding_window(
+        dataloader=test_dataload
+    )
+    print(test_loss,test_metric)
```
Comment (Muhammad-Al-Barbary) [inline tests/test_liver.py:288]: remove please

File: tests/test_liver.py [hunk 24]
```diff
@@ -0,0 +1,362 @@
...
+
+    assert round(test_loss,6) == 0.916104
+    assert round(test_metric.item(),4) == 0.0881
+
+
+
+
+def test_segment_liver(modality = "CT" , inference = "sliding_window", path = "tests/testdata/testcases/predicted_sliding.npy"):
```
Comment (Muhammad-Al-Barbary) [inline tests/test_liver.py:296]: This is the same test but with different parameters. Use parameterization.

**Review by Muhammad-Al-Barbary — APPROVED** [review]
A general comment is to always use parametrization whenever possible, clean code is not supposed to contain code duplication or hardcoding of values. 
Please address as many comments as you can before merging (after Dr. Hossam's approval).
